pkg/kv/kvserver/replica_raft.go

```diff
@@ -1718,9 +1718,13 @@ func (r *Replica) sendRaftMessages(
 				// Similar optimizations may be possible for other message types,
 				// although MsgAppResp is the only one that has been seen as a
 				// problem in practice.
-				if !message.Reject && message.Index > lastAppResp.Index {
-					lastAppResp = message
+				if !message.Reject {
 					drop = true
+					// NB: the first message "wins" the empty raftpb.Message{} here, so
+					// lastAppResp is always updated when we get here the first time.
+					if !msgAppRespBefore(&message, &lastAppResp) {
```

Comment:
Do we ever want to send more than one message here?

With #128497, we may want to extend the "happens before" to also take `Commit` index into account. For two responses with the same `Term` and `Index`, we should probably send one with the higher `Commit` index.

Reply:
One case to discuss here is when these messages span across terms and are addressed to different leaders. Maybe it's ok to send only the latest message to the latest leader? Or maybe we should send one (latest) message to each leader.